compiler/rustc_borrowck/src/diagnostics/region_errors.rs

```diff
@@ -548,21 +556,244 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
+    ///   --> $DIR/static-impl-obligation.rs:169:20
+    ///    |
+    /// LL |     impl dyn Foo + 'static where Self: 'static {
+    ///    |                    ^^^^^^^             ^^^^^^^
+    /// LL |         fn hello(&self) where Self: 'static {}
+    ///    |                                     ^^^^^^^
+    /// ```
+    #[allow(rustc::diagnostic_outside_of_impl)]
+    #[allow(rustc::untranslatable_diagnostic)]
+    fn explain_impl_static_obligation(
+        &self,
+        diag: &mut Diag<'_>,
+        ty: Ty<'tcx>,
+        call_span: Span,
+        outlived_fr: RegionVid,
+    ) {
+        let tcx = self.infcx.tcx;
+        let ty::FnDef(def_id, args) = ty.kind() else {
+            return;
+        };
+        let Ok(Some(instance)) = ty::Instance::resolve(
+            tcx,
+            self.param_env,
+            *def_id,
+            self.infcx.resolve_vars_if_possible(args),
+        ) else {
+            return;
+        };
+        debug!(?instance);
+        let def_id = instance.def_id();
+        let mut parent = tcx.parent(def_id);
+        debug!(?def_id, ?parent);
+        let trait_preds: Vec<_> = match tcx.def_kind(parent) {
+            hir::def::DefKind::Impl { .. } => tcx
+                .trait_id_of_impl(parent)
+                .map_or(vec![], |id| elaborate_predicates_of(tcx, id).collect()),
+            hir::def::DefKind::Trait => {
+                let Some(ty) = args.get(0).and_then(|arg| arg.as_type()) else {
+                    return;
+                };
+                let mut impls = vec![];
+                tcx.for_each_relevant_impl(parent, ty, |id| {
+                    impls.push(id);
+                });
+                if let [def_id] = impls[..] {
+                    // The method we have is on the trait, but for `parent` we want to analyze the
+                    // relevant impl instead.
+                    let preds = elaborate_predicates_of(tcx, parent);
+                    parent = def_id;
+                    preds.collect()
+                } else {
+                    return;
+                }
+            }
+            _ => return,
+        };
+        debug!(?def_id, ?parent);
+        let ty = tcx.type_of(parent).instantiate_identity();
+        debug!(?ty);
+        if self.to_error_region(outlived_fr) != Some(tcx.lifetimes.re_static) {
+            return;
+        }
+
+        // Look for `'static` bounds in the generics of the method and the `impl`.
+        // ```
+        // impl dyn Trait where Self: 'static {
+        //     fn foo(&self) where Self: 'static {}
+        // }
+        // ```
+        let mut predicates: Vec<Span> = elaborate_predicates_of(tcx, def_id)
+            .chain(elaborate_predicates_of(tcx, parent))
+            .chain(trait_preds)
+            .filter_map(filter_predicates(tcx.lifetimes.re_static, |pred_ty| {
+                self.infcx.can_eq(self.param_env, ty, pred_ty)
```

Comment:
This doesn't work if there are placeholder types, so this code will fail in weird and unnecessary ways for generic code. The predicates need to be instantiated with inference vars.